### PR TITLE
[Merged by Bors] - feat(Data/Finset/Sort): `sort (range n) = range n`

### PR DESCRIPTION
**Pull request by vihdzp**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

File: Mathlib/Data/Multiset/Sort.lean [hunk 1]
```diff
@@ -71,6 +71,9 @@ theorem sort_cons (a : α) (s : Multiset α) :
   refine Quot.inductionOn s fun l => ?_
   simpa [mergeSort_eq_insertionSort] using insertionSort_cons r (a := a) (l := l)
 
+theorem sort_range (n : ℕ) : sort (· ≤ ·) (range n) = List.range n :=
```
Comment (eric-wieser) [inline Mathlib/Data/Multiset/Sort.lean:74]: This should be `simp`

File: Mathlib/Data/Finset/Sort.lean [hunk 2]
```diff
@@ -76,6 +76,10 @@ theorem sort_insert [DecidableEq α] {a : α} {s : Finset α} (h₁ : ∀ b ∈
     sort r (insert a s) = a :: sort r s := by
   rw [← cons_eq_insert _ _ h₂, sort_cons r h₁]
 
+@[simp]
+theorem sort_range (n : ℕ) : sort (· ≤ ·) (range n) = List.range n :=
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Sort.lean:80]: Could you add the `<` version here and throughout too?

**Review by eric-wieser — APPROVED** [review]
Thanks, I think I had these in a branch somewhere.

It might be nice to have the ones about sorting in reverse order too, stated with `List.reverse`; up to you whether you want to declare those out of scope.

File: Mathlib/Data/Multiset/Sort.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/Data/Multiset/Sort.lean:74]: This doesn't work, as `simp` eagerly simplifies the LHS to `mergeSort (List.range n) = List.range n`. That feels like it shouldn't happen.

File: Mathlib/Data/Finset/Sort.lean (hunk 2 above)
Comment (vihdzp) [reply Mathlib/Data/Finset/Sort.lean:80]: The definition of `Multiset.sort` requires a total linear order, which `<` is not one of.

File: Mathlib/Data/Multiset/Sort.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/Data/Multiset/Sort.lean:74]: Actually, the issue might be the LHS is `List.range` rather than `Multiset.range`. I'll take a look.

**Comment by vihdzp** [comment]
I'm not sure what lemmas you're talking about. `sort (· ≤ ·) l.reverse = sort (· ≤ ·) l` and `(sort (· ≤ ·) l).reverse = sort (· ≥ ·) l` are both *false* in general, as e.g. if `x` and `y` are order equivalent, then `sort (· ≤ ·) [x, y] = sort (· ≥ ·) [x, y] = [x, y]` (recall merge-sort is a stable algorithm).

In any case, that seems quite out of scope for this PR.

File: Mathlib/Data/Multiset/Sort.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/Data/Multiset/Sort.lean:75]: This was indeed the issue, and has been fixed.

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**